[13.0][FIX] sale_order_line_packaging_qty

TDu opened the pull request:
When a quantity is set on the sale order line before or at the same time
than the product itself, the onchange triggered will replace the line
set quantity with the default size of the package.
This is especially problematic for sales order created programmatically.
So to improve the solution, if the quantity fits with the package
quantity it is kept.

simahawk committed:
sale_by_packaging: no pkging override if already set

Before this change the packaging was always overridden
hence the user had not control on the packaging to be used on SO line.

simahawk committed:
sale_by_packaging: skip auto assign when forcing uom qty

When the product_uom_qty is computed from inverse method of packaging qty
the packaging is already there and there's no need to auto-assign.

This also prevents skipping direct writes on prod uom qty or other keys
other than product_packaging.

TDu committed:
[13.0][FIX] sale_order_line_packaging_qty

When a quantity is set on the sale order line before or at the same time
than the product itself, the onchange triggered will replace the line
set quantity with the default size of the package.
This is especially problematic for sales order created programmatically.
So to improve the solution, if the quantity fits with the package
quantity it is kept.

lmignon commented on sale_order_line_packaging_qty/models/sale_order_line.py:
```diff
@@ -66,10 +66,16 @@ def _onchange_product_packaging_qty(self):
     @api.onchange("product_packaging")
     def _onchange_product_packaging(self):
         if self.product_packaging:
+            pack_qty = 1
```
@TDu Why don't you rely on the previous state of your record available provided by `self._origin`? You could therefore check `if self.product_uom_qty == self._origin.product_uom_qty:`

TDu replied:
@lmignon Because I did not think of it. So I guess you propose to do the check if existing product_uom_qty can be kept with the new packaging after the update (to save on some variable creation) ?
Well I did some tests on this but funny thing is, in my tests the `onchange` is not called following a change in a Form but by a `play_onchanges` function and there is no record in `self._origin` on every calll !?

lmignon replied:
@TDu I suppose that the play_onchanges is called on a record initialized with self.new()... If it's the case, to be consistent with odoo, the self.new() should be called with the origin args.... see https://github.com/odoo/odoo/blob/13.0/odoo/models.py#L5510.

TDu replied:
@lmignon Thanks for your answer, from what I can tell the sale.order is created with the `create` function. Done in an other OCA module https://github.com/OCA/edi/blob/cc04afc899246919a4ac90373d92ced0a398641c/sale_order_import/wizard/sale_order_import.py#L248

lmignon replied:
Something is missing into this puzzle... Nevertheless your proposed change makes sense even if I would prefer to not rely on some implicit rules to determine the context in which the onchange method is called. I'm curious to discover the method triggering the call to this onchange....

guewen replied:
`play_onchanges` comes from: https://github.com/OCA/server-tools/blob/5e432fdd9b33a18badff88119e9d8f0791736369/onchange_helper/models/base.py#L24

Is it possible that we have no `_origin` when going through this method?

lmignon replied:
@guewen Indeed, IMO, whe should initialize the "_origin" field on the current recordset before the call to `onchange` In this way we are consistent with the way the onchange methods are called from the UI. https://github.com/OCA/server-tools/blob/13.0/onchange_helper/models/base.py#L57
somethng like:

```python
    ....
     with self._set_origin():
         for field in onhcange_fields:
            onchange_values = self.onchange(all_values, field, onchange_specs)
    ...

@contextmanager
def _set_origin(self):
    if self:
        self.ensure_one()
        setattr(self, '_origin', self)
    yield
    if self:
        delattr(self, "_origin")
```

lmignon replied:
@guewen I'm not sure this code will work in odoo 13. The poposed PR is probably the safer way to solve your issue

lmignon replied:
Reading the code in models.py for odoo13. I'm sure that the proposed change will not work.

jbaudoux replied:
@TDu I tested with sale_order_import and the qty is wrong and/or the deducted packaging is wrong.

In my tests, I have a product with a retail box=22 units and transport box of 176 units.
Test 1: I order a qty not multiple of a packaging. Then SO line qty is always 1 retail box instead of the order qty
Test 2: I order a qty multiple of a transport box. Then SO line packaging is not expressed with the biggest packaging (transport box) but as a multiple of retail box.

When doing those 2 tests with odoo SO interface, it works properly

TDu replied:
@jbaudoux In the UI if the quantity is set first (with not a multiple of packaging quantity) and then the product is set, then the quantity is lost and 1 retail box is selected and the quantity contains the retail box quantity.

TDu replied:
This is what happens on the import. To change this behaviour would mean that a default package is not selected automatically. As by default Odoo sets 1 as a quantity I see no way to know if this is a default 1 or required 1 on the line.

jbaudoux replied:
@TDu I would remove the default packaging. It is not convenient in the UI as you do not see the possibilities without emptying the field. So better leave it empty by default and the user is free to select one. It will also be easier this way if the user does not want to select one as emptying that field is not that easy in the UI (it sometimes comes back).

TDu replied:
@jbaudoux I checked the code and the change happens in this onchange https://github.com/OCA/sale-workflow/blob/35feda502a35c55f0074232da1041926c0277902/sale_by_packaging/models/sale_order_line.py#L56
So only if the product is to be sold only by packaging. (This was the case for the product I used for my test).
Disabling the `Only sell by packaging`  on the product and the problem is gone.
So it is actually a feature of the module.

jbaudoux replied:
@TDu sell_only_by_packaging should require a packaging in the UI but not fill with default value

guewen replied:
What is the status here? Is there something to change or is this thread resolved?

TDu replied:
Yes by this https://github.com/OCA/sale-workflow/pull/1276/commits/6c291697db159c6c7d518526bae1eba43b9c7ccc

TDu replied:
It is a commit below

TDu replied:
The requirement is done by a constraint

Commit:
Fix test checking onchanges

The test was calling the onchanges manually to test their result, but as
some onchanges must be called in cascade to have a correct result, the
test tells more about itself, if it calls the onchanges in the correct
order or not.
Using Form will test them correctly in cascade automatically.

TDu committed:
Disable the setting of the packagin on product change

This feature was problematic because for order created programatically
By changing the packaging on product change the quantity ordered would
also change.
The correct packaging will be set on save.
And if not possible an error will be raised by a constraint.

i-vyshnevska commented on sale_by_packaging/models/sale_order_line.py:
```diff
@@ -52,18 +51,7 @@ def _check_product_packaging_sell_only_by_packaging(self):
-            if first_packaging:
-                self.update({"product_packaging": first_packaging.id})
-        return res
+        return super().product_id_change()
```
why not remove it completely?

TDu replied:
Yes why not ? Removed now.

TDu committed:
Fix linting and remove empty onchange

TDu commented on .pre-commit-config.yaml:
```diff
@@ -18,8 +18,8 @@ repos:
-  - repo: https://github.com/prettier/pre-commit
-    rev: "v1.19.1"
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: "v2.1.2"
```
This was to get the pre-commit hooks working again locally

guewen replied:
reference: https://github.com/OCA/oca-addons-repo-template/pull/41

yajo replied:
You can see here that prettier was working fine with the old config: https://travis-ci.com/github/OCA/sale-workflow/jobs/443358674#L284-L369

Could you please leave it as it was?

simahawk commented:
@TDu what's missing here?

TDu commented:
@simahawk Removing the change on the pre-commit config, done now.

simahawk committed:
sale_by_packaging: drop useless check on pkging multiple qty

This check is already done by core sale module.
Tests where not checkingt for that specific exception.

simahawk commented:
@TDu https://github.com/TDu/sale-workflow/pull/1 integrates https://github.com/OCA/sale-workflow/pull/1231

TDu committed:
sale_by_packaging: Fix tests and ask question

The product_product_packaging_qty on sale.order.line should always be an
integer, no ?

TDu committed:
sale_order_line_packaging_qty: prevent div by zero

simahawk commented on sale_by_packaging/tests/test_sale_only_by_packaging.py:
```diff
@@ -121,8 +66,11 @@ def test_create_auto_fill_packaging(self):
         )
-        self.assertFalse(order_line_1.product_packaging)
-        self.assertFalse(order_line_1.product_packaging_qty)
+        # Why should it be false if the quantity fits ?
```
@TDu is this something still needed?

TDu replied:
I kept it commented because, I did not understand the assertion.
For me it can be removed.

TDu replied:
removed

simahawk reviewed (approved):
LG

sebalix commented on sale_by_packaging/models/sale_order_line.py:
```diff
@@ -96,42 +78,26 @@ def _force_qty_with_package(self):
     def _onchange_product_uom_qty(self):
         self._force_qty_with_package()
         res = super()._onchange_product_uom_qty()
-        is_pack_multiple_warning = self._check_qty_is_pack_multiple()
-        if is_pack_multiple_warning:
-            self.product_packaging_qty = False
-            self.product_packaging = False
-        return res if res else is_pack_multiple_warning
-
-    def _check_qty_is_pack_multiple(self):
-        """ Check only for product with sell_only_by_packaging
-        """
-        # and we dont want to have this warning when we had the product
-        if self.product_id.sell_only_by_packaging:
-            if not self._get_product_packaging_having_multiple_qty(
-                self.product_id, self.product_uom_qty, self.product_uom
-            ):
-                warning_msg = {
-                    "title": _("Product quantity cannot be packed"),
-                    "message": _(
-                        "For the product {prod}\n"
-                        "The {qty} is not the multiple of any pack.\n"
-                        "Please add a package"
-                    ).format(prod=self.product_id.name, qty=self.product_uom_qty),
-                }
-                return {"warning": warning_msg}
-        return {}
```
As this code is removed, is the documentation needs to be updated?
https://github.com/OCA/sale-workflow/blob/13.0/sale_by_packaging/readme/DESCRIPTION.rst

TDu replied:
Readme has been updated.

TDu committed:
sale_by_packaging: update readme

simahawk commented:
@TDu pls squash

TDu commented:
Squashed, but only the fixup.

simahawk commented:
/ocabot merge patch
